### PR TITLE
Augmentation Documentation

### PR DESCRIPTION
**Pull request by cahuja1992**
Updated README with augmentation documentation. 
Usage of augmentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -418,6 +419,34 @@ python3 DeepSpeech.py --n_hidden 2048 --checkpoint_dir path/to/checkpoint/folder
...
 Note: the released models were trained with `--n_hidden 2048`, so you need to use that same value when initializing from the release models.
 
+### Training with augmentation
+
+Augmentation is a useful technique for better generalization of machine learning models. Thus, a pre-processing pipeline with various augmentation techniques on raw pcm and spectrogram has been implemented and can be used while training the model. 
+
+#### Audio Augmentation
+1. **Std for Gaussian additive noise:** ```data_aug_features_additive [0 - 1]```
+2. **Std for Normal distribution around 1 for multiplicative noise:** ```data_aug_features_multiplicative [0 - 1]``` 
+3. **Std for speeding-up tempo. If std is 0, this augmentation is not performed:** ```augmentation_speed_up_std [0 - 1]```
```
Comment (reuben) [inline README.md]: nit: In this list, expand "Std" to "Standard deviation".

Comment (reuben) [reply README.md]: Also, none of these flags have a range of [0, 1] so please remove that.

File: README.md [hunk 2]
```diff
@@ -418,6 +419,34 @@ python3 DeepSpeech.py --n_hidden 2048 --checkpoint_dir path/to/checkpoint/folder
...
+2. **Std for Normal distribution around 1 for multiplicative noise:** ```data_aug_features_multiplicative [0 - 1]``` 
+3. **Std for speeding-up tempo. If std is 0, this augmentation is not performed:** ```augmentation_speed_up_std [0 - 1]``` 
+
+#### Spectrogram Augmentation
+Inspired by Google Paper on [SpecAugment: A Simple Data Augmentation Method for Automatic Speech Recognition]( https://arxiv.org/abs/1904.08779)
+1. **Keep rate of dropout augmentation on a spectrogram (if 1, no dropout will be performed on the spectrogram)**: 
+   * Enable / Disable : ```augmentation_spec_dropout_keeprate [0 - 1]``` 
+
+2. **Whether to use frequency and time masking augmentation:** 
+   * Enable / Disable : ```augmentation_freq_and_time_masking [0, 1]```
```
Comment (reuben) [inline README.md]: This flag is a boolean, so it's `--augmentation_freq_and_time_masking` to enable and `--noaugmentation_freq_and_time_masking` to disable.

File: README.md [hunk 3]
```diff
@@ -418,6 +419,34 @@ python3 DeepSpeech.py --n_hidden 2048 --checkpoint_dir path/to/checkpoint/folder
...
+
+2. **Whether to use frequency and time masking augmentation:** 
+   * Enable / Disable : ```augmentation_freq_and_time_masking [0, 1]```  
+   * Max range of masks in the frequency domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_freq_mask_range eg: 5```
+   * Number of masks in the frequency domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_number_freq_masks eg: 3``` 
+   * Max range of masks in the time domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_time_mask_rangee eg: 2``` 
+   * Number of masks in the time domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_number_time_masks eg: 3 ``` 
+
+3. **Whether to use spectrogram speed and tempo scaling:** 
+   * Enable / Disable : ```augmentation_pitch_and_tempo_scaling [0,1]```
```
Comment (reuben) [inline README.md]: This flag is a boolean, so it's `--augmentation_pitch_and_tempo_scaling` to enable and `--noaugmentation_pitch_and_tempo_scaling` to disable.

File: README.md [hunk 4]
```diff
@@ -418,6 +419,34 @@ python3 DeepSpeech.py --n_hidden 2048 --checkpoint_dir path/to/checkpoint/folder
...
+   * Max range of masks in the frequency domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_freq_mask_range eg: 5```
+   * Number of masks in the frequency domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_number_freq_masks eg: 3``` 
+   * Max range of masks in the time domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_time_mask_rangee eg: 2``` 
+   * Number of masks in the time domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_number_time_masks eg: 3 ``` 
+
+3. **Whether to use spectrogram speed and tempo scaling:** 
+   * Enable / Disable : ```augmentation_pitch_and_tempo_scaling [0,1]```  
+   * Min value of pitch scaling: ```augmentation_pitch_and_tempo_scaling_min_pitch eg:0.95 ``` 
+   * Max value of pitch scaling: ```augmentation_pitch_and_tempo_scaling_max_pitch eg:1.2```  
+   * Max vlaue of tempo scaling: ```augmentation_pitch_and_tempo_scaling_max_tempo eg:1.2```
```
Comment (reuben) [inline README.md]: typo: vlaue

File: README.md [hunk 5]
```diff
@@ -71,6 +71,7 @@ See the output of `deepspeech -h` for more information on the use of `deepspeech
   - [Exporting a model for TFLite](#exporting-a-model-for-tflite)
   - [Making a mmap-able model for inference](#making-a-mmap-able-model-for-inference)
   - [Continuing training from a release model](#continuing-training-from-a-release-model)
+  - [Training with Augmentation](#training with augmentation)
```
Comment (reuben) [inline README.md]: Missing dashes in the link.

File: README.md [hunk 6]
```diff
@@ -421,30 +421,30 @@ Note: the released models were trained with `--n_hidden 2048`, so you need to us
 
 ### Training with augmentation
 
-Augmentation is a useful technique for better generalization of machine learning models. Thus, a pre-processing pipeline with various augmentation techniques on raw pcm and spectrogram has been implemented and can be used while training the model. 
+Augmentation is a useful technique for better generalization of machine learning models. Thus, a pre-processing pipeline with various augmentation techniques on raw pcm and spectrogram has been implemented and can be used while training the model. Following are the available augmentation techniques and can been used at the time of training while using these flags in command line options.
```
Comment (reuben) [inline README.md]: "and can been used at the time of training while using these flags in command line options." -> "that can be enabled at training time by using the corresponding flags in the command line."

File: README.md [hunk 7]
```diff
@@ -421,30 +421,30 @@ Note: the released models were trained with `--n_hidden 2048`, so you need to us
...
    * Max range of masks in the time domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_time_mask_rangee eg: 2``` 
    * Number of masks in the time domain when performing freqtime-mask augmentation: ```augmentation_freq_and_time_masking_number_time_masks eg: 3 ``` 
 
 3. **Whether to use spectrogram speed and tempo scaling:** 
-   * Enable / Disable : ```augmentation_pitch_and_tempo_scaling [0,1]```  
+   * Enable / Disable : ```a--augmentation_pitch_and_tempo_scaling to enable and --noaugmentation_pitch_and_tempo_scaling to disable.```  
    * Min value of pitch scaling: ```augmentation_pitch_and_tempo_scaling_min_pitch eg:0.95 ``` 
    * Max value of pitch scaling: ```augmentation_pitch_and_tempo_scaling_max_pitch eg:1.2```  
-   * Max vlaue of tempo scaling: ```augmentation_pitch_and_tempo_scaling_max_tempo eg:1.2```  
+   * Max valaue of tempo scaling: ```augmentation_pitch_and_tempo_scaling_max_tempo eg:1.2```
```
Comment (reuben) [inline README.md]: Looks like you added another typo in the process of fixing the previous typo :P

**Comment by rhamnett** [comment]
Hello, it's difficult to understand what a practical value for the std value of flags such as data_aug_features_additive/data_aug_features_multiplicative without a more detailed explanation or example.

For example, what is the result of the noise by setting it closer to 1 than closer to 0?

**Comment by bernardohenz** [comment]
@rhamnett both augs on features (additive and multiplicative noise) follow a normal distribution with the indicated standard deviation. If you set the standard deviation of ```data_aug_features_additive``` of 0.5, it means that 99.7% of values that will be added to the features will be within the range [-1.5, 1.5] (which corresponds to 3 stds around the mean). The same goes for multiplicative (but in this case the mean is set to 1, as this the neutral factor of multiplication).

You can check more information about normal distribution in the following links: [std_deviation](https://en.wikipedia.org/wiki/Standard_deviation), [sigma_rule](https://en.wikipedia.org/wiki/68–95–99.7_rule).

Now, to which range of values is the best, I'm afraid you can only find by experimenting different setups. You can try to check/debug which values are normally found on feature-space and try to get some intuition on how to set these parameters, but in the end you will have to try them to check which parameters work best.

**Pull request merged**